ostd/src/arch/riscv/pci.rs

```diff
@@ -1,8 +1,60 @@
 // SPDX-License-Identifier: MPL-2.0
 
-//! PCI bus io port
+//! PCI bus access
 
-use super::device::io_port::{IoPort, ReadWriteAccess, WriteOnlyAccess};
+use spin::Once;
 
-pub static PCI_ADDRESS_PORT: IoPort<u32, WriteOnlyAccess> = unsafe { IoPort::new(0x0) };
-pub static PCI_DATA_PORT: IoPort<u32, ReadWriteAccess> = unsafe { IoPort::new(0x0) };
+use super::boot::DEVICE_TREE;
+use crate::{bus::pci::PciDeviceLocation, io::IoMem, mm::VmIoOnce, prelude::*, Error};
+
+static PCI_BASE_ADDR: Once<IoMem> = Once::new();
+
+pub(crate) fn write32(location: &PciDeviceLocation, offset: u32, value: u32) -> Result<()> {
+    PCI_BASE_ADDR.get().ok_or(Error::IoError)?.write_once(
+        (encode_as_address_offset(location) | (offset & 0xfc)) as usize,
+        &value,
+    )
+}
+
+pub(crate) fn read32(location: &PciDeviceLocation, offset: u32) -> Result<u32> {
+    PCI_BASE_ADDR
+        .get()
+        .ok_or(Error::IoError)?
+        .read_once((encode_as_address_offset(location) | (offset & 0xfc)) as usize)
+}
+
+pub(crate) fn has_pci_bus() -> bool {
+    PCI_BASE_ADDR.is_completed()
+}
+
+pub(crate) fn init() -> Result<()> {
+    let pci = DEVICE_TREE
+        .get()
+        .unwrap()
+        .find_node("/soc/pci")
+        .ok_or(Error::IoError)?;
+
+    let mut reg = pci.reg().ok_or(Error::IoError)?;
+    let region = reg.next().ok_or(Error::IoError)?;
+    PCI_BASE_ADDR.call_once(|| {
+        IoMem::acquire(
+            (region.starting_address as usize)
+                ..(region.starting_address as usize + region.size.unwrap()),
+        )
+        .unwrap()
+    });
+
+    if reg.count() != 0 {
+        warn!("PCI node should have exactly one reg property");
+        return Err(Error::IoError);
+    }
```

Comment:
Since you cannot figure out in what scenario the PCI node can contain 2 `reg`s, how about:
```rust
    let mut reg = pci.reg().ok_or(Error::IoError)?;
    if reg.len() != 1 {
        warn!("PCI node should have exactly one `reg` property, but found {} `reg`s", reg.len());
        return Err(Error::IoError);
    }

    let region = reg[0];
    PCI_BASE_ADDR.call_once(|| {
        IoMem::acquire(
            (region.starting_address as usize)
                ..(region.starting_address as usize + region.size.unwrap()),
        )
        .unwrap()
    });
```